File: math/test/LDIndexLineListTest.cpp
/**
 *
 *  2015/5/24
 *	aso
 *
 *  Copyright(c) Live2D Inc. All rights reserved.
 *  [[ CONFIDENTIAL ]]
 */

#include "LDIndexLineListTest.h"

#include <QTest>
#include <QDebug>
#include "../serialize/SerializeHelper.h"
#include "LDIndexLineList.h"
#include "LDSerialize.h"
namespace live2d {

using namespace serialize;


void LDIndexLineListTest::serializeTest()
{
	const char* name="lineList";
	QSharedPointer<LDIndexLineList> src(new LDIndexLineList);
	src->push_back(LDIndexLine(0,1));

	//シリアライズ
	SerializeHelper::writeBoostXml( name,src );

	//デシリアライズ
	auto dst=SerializeHelper::readBoostXml<LDIndexLineList>( name );

	QVERIFY(*(dst.data())==*(src.data()));
}

void LDIndexLineListTest::hitTest()
{
	LDIndexLineList lines;
	LDIndexLine line_1(0,1);
	LDIndexLine line_2(1,2);
	LDIndexLine line_3(2,3);
	LDIndexLine line_4(3,0);
	lines<<line_1;
	lines<<line_2;
	lines<<line_3;
	lines<<line_4;

	LDPointList points;
	points.push_back(LDPoint(0,0));
	points.push_back(LDPoint(100,0));
	points.push_back(LDPoint(100,100));
	points.push_back(LDPoint(0,100));
	
	
	QCOMPARE(lines.isHit(points, LDPoint(0,0),1),true);//頂点上
	QCOMPARE(lines.isHit(points, LDPoint(50,0),1),true);//線上
	QCOMPARE(lines.isHit(points, LDPoint(50,1),3),true);//線からちょっと離れたところ
	QCOMPARE(lines.isHit(points, LDPoint(50,4),3),false);//線から範囲外に離れたところ
	QCOMPARE(lines.isHit(points, LDPoint(200,0),3),false);//線の延長戦上に離れたところ
	QCOMPARE(lines.isHit(points, LDPoint(200,1),3),false);//線の延長戦上に離れたところ
	
	
	
	
	
}

void LDIndexLineListTest::constructTest()
{
	QList<int> int_list;
	int_list.push_back(0);
	int_list.push_back(1);
	int_list.push_back(2);	
	int_list.push_back(3);
//	int_list.push_back(0);
	
	
	LDIndexLineList list(int_list);
	
	
	
	LDIndexLineList compare;
	compare.push_back(LDIndexLine(0,1));
	compare.push_back(LDIndexLine(1,2));
	compare.push_back(LDIndexLine(2,3));
	compare.push_back(LDIndexLine(3,0));
	
	QCOMPARE(list,compare);
	
	
}

} // namespace live2d

File: cs/ConsoleApplication1/LDIndexLineListTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace live2d
{
    public class LDIndexLineListTest : ITest
    {
        public void test()
        {
            this.constructTest();
            this.hitTest();
        }

        /*
        public void serializeTest()
        {
            const char* name = "lineList";
            QSharedPointer<LDIndexLineList> src(new LDIndexLineList);
            src->push_back(LDIndexLine(0, 1));

            //シリアライズ
            SerializeHelper::writeBoostXml(name, src);

            //デシリアライズ
            auto dst = SerializeHelper::readBoostXml<LDIndexLineList>(name);

            QVERIFY(*(dst.data()) == *(src.data()));
        }
        */

        public void hitTest()
        {
            LDIndexLineList lines = new LDIndexLineList();
            LDIndexLine line_1 = new LDIndexLine(0, 1);
            LDIndexLine line_2 = new LDIndexLine(1, 2);
            LDIndexLine line_3 = new LDIndexLine(2, 3);
            LDIndexLine line_4 = new LDIndexLine(3, 0);
            lines.add(line_1);
            lines.add(line_2);
            lines.add(line_3);
            lines.add(line_4);

            LDPointList points = new LDPointList();
            points.Add(new LDPoint(0, 0));
            points.Add(new LDPoint(100, 0));
            points.Add(new LDPoint(100, 100));
            points.Add(new LDPoint(0, 100));


            TestUtil.COMPARE(lines.isHit(points, new LDPoint(0, 0), 1), true);//頂点上
            TestUtil.COMPARE(lines.isHit(points, new LDPoint(50, 0), 1), true);//線上
            TestUtil.COMPARE(lines.isHit(points, new LDPoint(50, 1), 3), true);//線からちょっと離れたところ
            TestUtil.COMPARE(lines.isHit(points, new LDPoint(50, 4), 3), false);//線から範囲外に離れたところ
            TestUtil.COMPARE(lines.isHit(points, new LDPoint(200, 0), 3), false);//線の延長戦上に離れたところ
            TestUtil.COMPARE(lines.isHit(points, new LDPoint(200, 1), 3), false);//線の延長戦上に離れたところ





        }

        public void constructTest()
        {
            List<int> int_list = new List<int>();
            int_list.Add(0);
            int_list.Add(1);
            int_list.Add(2);
            int_list.Add(3);
            //	int_list.push_back(0);
            LDIndexLineList list = new LDIndexLineList(int_list);

            LDIndexLineList compare = new LDIndexLineList();
            compare.Add(new LDIndexLine(0, 1));
            compare.Add(new LDIndexLine(1, 2));
            compare.Add(new LDIndexLine(2, 3));
            compare.Add(new LDIndexLine(3, 0));
            TestUtil.COMPARE(list, compare);
        }
    }
}
